Adding a visibility toggle for FTextField.password (#668)

MoathCodes opened the pull request:
**Describe the changes**
This PR is from #668, these changes add new visibility toggle for the `FTextField.password`, I still need to change the `FTextFormField.password` as well, **however while implementing this I realized that because I've made the TextField into a static method the parameters no longer have documentations**.
I think I've missed some better architectural decisions, please tell me what do you think.
###  Usage Examples

```dart
// ✅ Basic usage - automatic toggle and controller
FTextField.password()

// ✅ Custom toggle - requires explicit controller
final controller = ValueNotifier<bool>(false);
FTextField.password(
  visibilityController: controller,
  toggleBuilder: (context, style, states) => IconButton(
    onPressed: () => controller.value = !controller.value,
    icon: Icon(controller.value ? Icons.visibility_off : Icons.visibility),
  ),
)

// ✅ No toggle - password always obscured  
FTextField.password(toggleBuilder: null)
```
**Checklist**
- [x] I have read the [CONTRIBUTING.md](../CONTRIBUTING.md).
- [ ] I have included the relevant unit/golden tests.
- [ ] I have included the relevant samples.
- [x] I have updated the documentation accordingly.
- [ ] I have added the required flutters_hook for widget controllers.
- [ ] I have updated the [CHANGELOG.md](../forui/CHANGELOG.md) if necessary.

Pante commented on forui/lib/src/widgets/text_field/text_field.dart:
```diff
@@ -37,6 +37,17 @@ typedef FFieldBuilder<T> = Widget Function(BuildContext context, T style, Set<Wi
 /// See [FTextField.prefixBuilder] and [FTextField.suffixBuilder].
 typedef FFieldIconBuilder<T> = Widget Function(BuildContext context, T style, Set<WidgetState> states);
 
+/// A callback for building a password visibility toggle button.
+///
+/// [style] is the field's style.
+/// [states] is the current states of the widget.
+///
+/// The builder should use the provided `visibilityController` to toggle visibility
+/// and listen to visibility state changes.
+///
+/// See [FTextField.password].
+typedef FPasswordToggleBuilder = Widget Function(BuildContext context, FTextFieldStyle style, Set<WidgetState> states);
+
```
I think we can re-use `FFieldIconBuilder` instead of declaring a new builder.

Pante commented on forui/lib/src/widgets/text_field/text_field.dart:
````diff
@@ -983,6 +920,194 @@ class FTextField extends StatefulWidget {
+  ///   toggleBuilder: null,
+  /// )
+  /// ```
+  static Widget password({
````
I think we can create a private subclass of FTextField. That way we can use a redirecting factory constructor thereby solving this issue.

Pante commented on forui/lib/src/widgets/text_field/text_field.dart:
```diff
@@ -983,6 +920,194 @@ class FTextField extends StatefulWidget {
+    bool Function(TextEditingValue) clearable = Defaults.clearable,
+    String? initialText,
+    FPasswordToggleBuilder? toggleBuilder = _defaultToggleBuilder,
+    FValueNotifier<bool>? visibilityController,
```
I think we can use a plain `ValueNotifier` here. Also I think we should rename this to `obscuredText`, kind of treating the notifier as a overload of the original `obscuredText`.

Pante commented on forui/lib/src/widgets/text_field/text_field.dart:
```diff
@@ -1062,6 +1187,359 @@ class FTextField extends StatefulWidget {
+  Widget build(BuildContext context) => ValueListenableBuilder<bool>(
+    valueListenable: _visibilityController,
+    builder: (context, isVisible, child) {
+      // Determine the suffix builder based on priority:
+      // 1. Custom suffixBuilder (highest priority - disables toggle)
+      // 2. toggleBuilder (if not null, shows toggle using internal controller for default, requires external controller for custom)
+      // 3. null (no suffix)
+      final FFieldIconBuilder<FTextFieldStyle>? effectiveSuffixBuilder =
+          widget.suffixBuilder ??
+          (widget.toggleBuilder != null
+              ? (context, style, states) => _buildToggleWidget(context, style, states, isVisible)
+              : null);
```
I think we should either:
* Automatically wrap these in a row. 
* Change the API to accept either a suffix builder or toggle builder, not both.

Pante reviewed (commented):
Sorry for the late reply.

MoathCodes commented:
I've removed the toggle builder so now it just uses the `suffixBuilder`, and the toggle can be disabled by just `suffixBuilder: null`, I've also changed the implementation into a private subclass as you've suggested, but I still couldn't find a way to make the parameters have documentations.

MoathCodes commented:
> Generally LGTM, just a few minor nits (Remember to update the README.md too).

I'm sorry which README.md file exactly?

Pante commented:
My bad, CHANGELOG.md not README.md*

MoathCodes commented:
It's my first time dealing with golden test hopefully I've done it right.

Another thing I see that the FTextFormField.password can be simplified if we use the FTextField.password but I couldn't find a way because i kept getting into problems with the `_FormField` class, I was afraid that I could break the form state I had a couple of ideas like passing an optional builder to the _FormField class that if not null will build FTextField.password otherwise it behaves as is, What do you think?

Pante reviewed (approved):
LGTM! Thanks for the PR!

MoathCodes commented:
Happy to help with this amazing project!

Pante commented:
I'm probably going to merge this PR later during the weekends. Just so I remember to do the remaining stuff.

MoathCodes commented:
I just realized that I've committed the edits to the sandbox file, I was testing and forgot to restore the file sorry.
